File: TeamCode/src/main/java/org/firstinspires/ftc/teamcode/Subsystems/Swerve.java
package org.firstinspires.ftc.teamcode.Subsystems;

import com.qualcomm.hardware.rev.RevHubOrientationOnRobot;
import com.qualcomm.robotcore.hardware.HardwareMap;
import com.qualcomm.robotcore.hardware.IMU;

import org.firstinspires.ftc.robotcore.external.navigation.AngleUnit;
import org.firstinspires.ftc.teamcode.Constants;
import org.firstinspires.ftc.teamcode.Utilities.EZTelemetry;

public class Swerve {

    private final EZTelemetry telem;

    private final SwerveModule[] mods = new SwerveModule[4];
    private FusionOdometry odom = null;
    private IMU imu = null;

    // max theoretical speed is 1.93m/s
    // max theoretical angular velocity is ~5.20884 rad/s
    private double MAX_SPEED_MPS = 1.93;
    private final double MAX_ANGULAR_VELOCITY_RAD_PER_SECONDS = 5.20884002936;

    private boolean enableTelemetry;

    protected double lastHeading = 0;


    public Swerve(HardwareMap hardwareMap, EZTelemetry telem) {

        this.telem = telem;

        mods[0] = new SwerveModule(hardwareMap, telem, Constants.DriveTrainConstants.Mod0.modConstants); // front-left
        mods[1] = new SwerveModule(hardwareMap, telem, Constants.DriveTrainConstants.Mod1.modConstants); // front-right
        mods[2] = new SwerveModule(hardwareMap, telem, Constants.DriveTrainConstants.Mod2.modConstants); // back-left
        mods[3] = new SwerveModule(hardwareMap, telem, Constants.DriveTrainConstants.Mod3.modConstants); // back-right

        RevHubOrientationOnRobot.LogoFacingDirection logo = RevHubOrientationOnRobot.LogoFacingDirection.UP;
        RevHubOrientationOnRobot.UsbFacingDirection usb = RevHubOrientationOnRobot.UsbFacingDirection.RIGHT;
        RevHubOrientationOnRobot orientationOnRobot = new RevHubOrientationOnRobot(logo, usb);

        imu = hardwareMap.get(IMU.class, "imu");
        imu.initialize(new IMU.Parameters(orientationOnRobot));

        enableTelemetry = false;

        double halfTrack = Constants.DriveTrainConstants.trackWidth / 2.0;
        double halfWheelbase = Constants.DriveTrainConstants.wheelbase / 2.0;

    }

    public Swerve(HardwareMap hardwareMap, EZTelemetry telem, FusionOdometry odometry) {

        this.telem = telem;
        this.odom = odometry;

        mods[0] = new SwerveModule(hardwareMap, telem, Constants.DriveTrainConstants.Mod0.modConstants); // front-left
        mods[1] = new SwerveModule(hardwareMap, telem, Constants.DriveTrainConstants.Mod1.modConstants); // front-right
        mods[2] = new SwerveModule(hardwareMap, telem, Constants.DriveTrainConstants.Mod2.modConstants); // back-left
        mods[3] = new SwerveModule(hardwareMap, telem, Constants.DriveTrainConstants.Mod3.modConstants); // back-right

        enableTelemetry = false;
    }

    public void stop() {
        for (int i = 0; i < 4; i++) {
            mods[i].setDrivePower(0);
            mods[i].setModuleSetpoint(mods[i].getDegrees(true));
            mods[i].setTurnSpeed(0);
        }
    }

    public void drive(double xVal, double yVal, double rVal, boolean fieldRelative) {

        if (Math.abs(xVal) < 0.03) xVal = 0;
        if (Math.abs(yVal) < 0.03) yVal = 0;
        if (Math.abs(rVal) < 0.03) rVal = 0;

        double x = yVal;
        double y = -xVal;

        if (fieldRelative) {
            double headingRad = Math.toRadians(getHeading());
            double cos = Math.cos(headingRad);
            double sien = Math.sin(headingRad);
            double xTemp = x;
            double yTemp = y;

            x = xTemp * cos + yTemp * sien;
            y = -xTemp * sien + yTemp * cos;
        }

        double r = -rVal;

        final double rotX = r * (Constants.DriveTrainConstants.trackWidth / Constants.DriveTrainConstants.moduleHypotenuse);
        final double rotY = r * (Constants.DriveTrainConstants.wheelbase / Constants.DriveTrainConstants.moduleHypotenuse);

        final double xLeft = x - rotX;
        final double xRight = x + rotX;
        final double yBack = y - rotY;
        final double yFront = y + rotY;

        double mod0Speed = Math.hypot(xLeft, yFront);
        double mod1Speed = Math.hypot(xRight, yFront);
        double mod2Speed = Math.hypot(xLeft, yBack);
        double mod3Speed = Math.hypot(xRight, yBack);

        double max = mod0Speed;
        if (mod1Speed > max) max = mod1Speed;
        if (mod2Speed > max) max = mod2Speed;
        if (mod3Speed > max) max = mod3Speed;

        if (max > 1.0) {
            double inv = 1.0 / max;
            mod0Speed *= inv; mod1Speed *= inv; mod2Speed *= inv; mod3Speed *= inv;
        }

        double a0 = Math.toDegrees(Math.atan2(yFront, xLeft));
        if (a0 < 0) a0 += 360.0;
        double a1 = Math.toDegrees(Math.atan2(yFront, xRight));
        if (a1 < 0) a1 += 360.0;
        double a2 = Math.toDegrees(Math.atan2(yBack, xLeft));
        if (a2 < 0) a2 += 360.0;
        double a3 = Math.toDegrees(Math.atan2(yBack, xRight));
        if (a3 < 0) a3 += 360.0;

        mods[0].setDrivePower(mod0Speed);
        mods[1].setDrivePower(mod1Speed);
        mods[2].setDrivePower(mod2Speed);
        mods[3].setDrivePower(mod3Speed);

        if (xVal != 0 || yVal != 0 || rVal != 0) {
            mods[0].setModuleSetpoint(a0);
            mods[1].setModuleSetpoint(a1);
            mods[2].setModuleSetpoint(a2);
            mods[3].setModuleSetpoint(a3);
        }

        mods[0].setModulePosition();
        mods[1].setModulePosition();
        mods[2].setModulePosition();
        mods[3].setModulePosition();
    }

    public void drivePrep(double xVal, double yVal, double rVal, boolean fieldRelative) {

        if (Math.abs(xVal) < 0.03) xVal = 0;
        if (Math.abs(yVal) < 0.03) yVal = 0;
        if (Math.abs(rVal) < 0.03) rVal = 0;

        double x = yVal;
        double y = -xVal;

        if (fieldRelative) {
            double headingRad = Math.toRadians(getHeading());
            double cos = Math.cos(headingRad);
            double sien = Math.sin(headingRad);
            double xTemp = x;
            double yTemp = y;

            x = xTemp * cos + yTemp * sien;
            y = -xTemp * sien + yTemp * cos;
        }

        double r = -rVal;

        final double rotX = r * (Constants.DriveTrainConstants.trackWidth / Constants.DriveTrainConstants.moduleHypotenuse);
        final double rotY = r * (Constants.DriveTrainConstants.wheelbase / Constants.DriveTrainConstants.moduleHypotenuse);

        final double xLeft = x - rotX;
        final double xRight = x + rotX;
        final double yBack = y - rotY;
        final double yFront = y + rotY;

        double a0 = Math.toDegrees(Math.atan2(yFront, xLeft));
        if (a0 < 0) a0 += 360.0;
        double a1 = Math.toDegrees(Math.atan2(yFront, xRight));
        if (a1 < 0) a1 += 360.0;
        double a2 = Math.toDegrees(Math.atan2(yBack, xLeft));
        if (a2 < 0) a2 += 360.0;
        double a3 = Math.toDegrees(Math.atan2(yBack, xRight));
        if (a3 < 0) a3 += 360.0;

        mods[0].setDrivePower(0);
        mods[1].setDrivePower(0);
        mods[2].setDrivePower(0);
        mods[3].setDrivePower(0);

        if (xVal != 0 || yVal != 0 || rVal != 0) {
            mods[0].setModuleSetpoint(a0);
            mods[1].setModuleSetpoint(a1);
            mods[2].setModuleSetpoint(a2);
            mods[3].setModuleSetpoint(a3);
        }

        mods[0].setModulePosition();
        mods[1].setModulePosition();
        mods[2].setModulePosition();
        mods[3].setModulePosition();
    }

    public double getHeading() {
        if(imu != null ){
            double rotation = imu.getRobotYawPitchRollAngles().getYaw(AngleUnit.DEGREES);
            rotation += 360.0;
            rotation %= 360.0;
            return rotation;
        } else {
            return odom.getHeading();
        }
    }

    public void zeroGyro() {
        if(imu != null) {
            imu.resetYaw();
        } else {
            odom.zeroGyro();
        }
    }

    public void toggleTelemetry() {
        enableTelemetry = !enableTelemetry;
    }

    public void skadoodle() {
    }
}

File: TeamCode/src/main/java/org/firstinspires/ftc/teamcode/Subsystems/Swerve.kt
package org.firstinspires.ftc.teamcode.Subsystems

import com.qualcomm.robotcore.hardware.HardwareMap
import org.firstinspires.ftc.robotcore.external.Telemetry
import org.firstinspires.ftc.teamcode.Constants
import kotlin.math.abs
import kotlin.math.atan2
import kotlin.math.cos
import kotlin.math.hypot
import kotlin.math.max
import kotlin.math.sin

class Swerve(hardwareMap: HardwareMap, telemetry: Telemetry) {
    private val telemetry: Telemetry?

    private val mod0: SwerveModule
    private val mod1: SwerveModule
    private val mod2: SwerveModule
    private val mod3: SwerveModule

    private val otos: OTOSSensor

    init {
        this.telemetry = telemetry

        mod0 = SwerveModule(hardwareMap, telemetry, Constants.DriveTrainConstants.Mod0.modConstants)
        mod1 = SwerveModule(hardwareMap, telemetry, Constants.DriveTrainConstants.Mod1.modConstants)
        mod2 = SwerveModule(hardwareMap, telemetry, Constants.DriveTrainConstants.Mod2.modConstants)
        mod3 = SwerveModule(hardwareMap, telemetry, Constants.DriveTrainConstants.Mod3.modConstants)

        otos = OTOSSensor(hardwareMap, telemetry)
    }

    fun drive(xVal: Double, yVal: Double, rVal: Double, fieldRelative: Boolean) {
        var x = xVal
        var y = yVal
        var r = rVal

        if (abs(x) < 0.1) {
            x = 0.0
        }
        if (abs(y) < 0.1) {
            y = 0.0
        }
        if (abs(r) < 0.1) {
            r = 0.0
        }

        if (fieldRelative) {
            val robotHeading = Math.toRadians(otos.heading)

            x = x * cos(robotHeading) + y * sin(robotHeading)
            y = -x * sin(robotHeading) + y * cos(robotHeading)
        }

        val widthVector = r * Constants.DriveTrainConstants.widthRotation
        val lengthVector = r * Constants.DriveTrainConstants.lengthRotation

        val aVector = x - widthVector
        val bVector = x + widthVector
        val cVector = y - lengthVector
        val dVector = y + lengthVector

        var mod0Speed = hypot(bVector, dVector)
        var mod1Speed = hypot(bVector, cVector)
        var mod2Speed = hypot(aVector, dVector)
        var mod3Speed = hypot(aVector, cVector)

        var max = max(abs(mod0Speed), abs(mod1Speed))
        max = max(max, abs(mod2Speed))
        max = max(max, abs(mod3Speed))

        if (max > 1.0) {
            mod0Speed /= max
            mod1Speed /= max
            mod2Speed /= max
            mod3Speed /= max
        }

        var mod0Angle = Math.toDegrees(atan2(-bVector, dVector))
        var mod1Angle = Math.toDegrees(atan2(-bVector, cVector))
        var mod2Angle = Math.toDegrees(atan2(-aVector, dVector))
        var mod3Angle = Math.toDegrees(atan2(-aVector, cVector))
        mod0Angle = (mod0Angle + 360) % 360
        mod1Angle = (mod1Angle + 360) % 360
        mod2Angle = (mod2Angle + 360) % 360
        mod3Angle = (mod3Angle + 360) % 360

        mod0.setDrivePower(mod0Speed)
        mod1.setDrivePower(mod1Speed)
        mod2.setDrivePower(mod2Speed)
        mod3.setDrivePower(mod3Speed)

        if (abs(xVal) > 0.1 || abs(yVal) > 0.1 || abs(rVal) > 0.1) {
            mod0.setModuleSetpoint(mod0Angle)
            mod1.setModuleSetpoint(mod1Angle)
            mod2.setModuleSetpoint(mod2Angle)
            mod3.setModuleSetpoint(mod3Angle)
        }

        mod0.setModulePosition()
        mod1.setModulePosition()
        mod2.setModulePosition()
        mod3.setModulePosition()
    }

    fun update() {
        telemetry!!.addLine("Swerve")
        telemetry.addData("X Position ", otos.pose?.x)
        telemetry.addData("Y Position ", otos.pose?.y)
        telemetry.addData("Heading ", otos.heading)
        telemetry.addData("OTOS Heading ", otos.pose?.h)
        telemetry.addLine()
    }
}
